### ISSUE TITLE
Missing glyph

### ISSUE THREAD

**Issue by lemzwerg** [pts/pdfsizeopt#11]
[pts/pdfsizeopt@5a1e317a]
[pts/pdfsizeopt-jbig2@26cade6b]
[pts/sam2p@c5c3b1cb]
[pts/tif22pnm@8abf77d7]
[pngout Mar 19 2015]
[Multivalent.jar downloaded April 2010]
[gs 9.20]

[This large PDF file](https://github.com/pts/pdfsizeopt/files/860455/notation.zip), generated by a recent lilypond developer version, seems to [convert fine](https://github.com/pts/pdfsizeopt/files/860481/notation.psom.pdf) (BTW, this works much better than testing pdfsizeopt a few years ago, which then created far more artifacts), except for one font glitch.  After conversion with
```
pdfsizeopt.py --use-pngout=true --use-jbig2=true --use-multivalent=true notation.pdf
```
there are square boxes instead of asterisks for the pedal end markers, cf. logical pages 13 and 680 of the document (= numbered pages 1 and 668).

**Comment by pts** [comment]
Thank you for reporting this bug!

Most of the font processing of `pdfsizeopt` is done by `gs`, and it's sometimes buggy. The correct fix would be using another library (e.g. `libpoppler`) instead of `gs` or rewriting the font processing code from scratch. I'm unlikely to work on it soon, unless my motivation changes significantly.

In the meantime, you can specify `--do-unify-fonts=false`, and then `--do-optimize-fonts=false` as a workaround. Please let me know which of those produces correct output.

`--use-pngout=...` and `--use-jbig2=...` are unlikely to make a difference here.

`--use-multivalent=false` may make a difference, it's worth trying.

Please note that https://github.com/pts/pdfsizeopt now has new, simplified installation instructions; you don't need to download jbig2, sam2p, tif22pnm, pngout and gs separately anymore on Linux, FreeBSD and Windows.

**Comment by lemzwerg** [comment]
Using `--do-unify-fonts=false` indeed fixes this issue!  It would be great if you could report the problem to the GS people so that it gets eventually fixed...

I've seen the simplified installation instructions, thanks.  However, I wanted to use the git repositories directly, so I followed the instructions given for generic Unix later in the `README.txt` file.

**Comment by pts** [comment]
Even though I've made some improvements recently to the font unification implementation (version 2f063aac383021b4faacd22a5e5918cb43454a9b), this problem has not been fixed yet: a rectangle still appears on physical page 13 if `--do-unify-fonts=false` is not specified.

**Comment by pts** [comment]
I was able to reproduce this problem with a much smaller PDF file: [emmentaler20.pdf](https://github.com/pts/pdfsizeopt/files/1072129/emmentaler20.pdf). It looks like the problem is with the conversion of the *Emmentaler-20* font by Ghostscript initiated by pdfsizeopt. What exactly breaks here need further investigation.

As a workaround until this is fixed, please use `pdfsizeopt --do-unify-fonts=false --do-regenerate-all-fonts=false`.

**Comment by pts** [comment]
An even smaller input file to reproduce the bug with: [emmentaler20b.pdf](https://github.com/pts/pdfsizeopt/files/1074525/emmentaler20b.pdf)

**Comment by pts** [comment]
I was able to fix this in 7ff3bfa5dc2643cd4ba9daf7f1fe1e9613be112c. Please upgrade to the latest pdfsizeopt, and enjoy correct output for all glyphs in notation.pdf.

Here is some explanation on what was wrong and how I fixed it.

Only the `/pedal.*` glyph is broken in the font *Emmentaler-20* when being converted by pdfsizeopt. That's because the special treatment of the `#` escaping in PDF name objects:

* If the `/Type/Encoding` `/Differences` contains `/pedal.*`, then the Type1C font stream should contain `pedal.*`. (This is the case in in the input notation.pdf .)
* If the `/Type/Encoding` `/Differences` contains `/pedal.#2A`, then the Type1C font stream should contain `pedal.*`. (This is the case in the inputs emmentaler20.pdf and emmentaler20b.pdf.)
* If the `/Type/Encoding` `/Differences` contains `/pedal.#232A`, then the Type1C font stream should contain `pedal.#2A`.

pdfsizeopt's *OptimizeType1CFonts* function had a bug: it generated a mismatching glyph name in `/Differences` and in the Type1C font stream. (It looks like Ghostscript was working correctly.) *ParseTyep1CFonts* returned glyph names escaped (i.e. if the font contained `pedal.*`, then it returned `/pedal.#2A`), and when regenerating the font data, *AppendSerialized* didn't do the corresponding unescaping. The unescaping was needed, because Type1CGenerator expected PostScript input (and there is no automatic name unescaping there, only PDF has it). The fix in 7ff3bfa5dc2643cd4ba9daf7f1fe1e9613be112c added this unescaping.

**Comment by lemzwerg** [comment]
Thanks a lot!  Will test soon and report any further problems :-)
